### ISSUE TITLE
SSC: Support Passing In `initialValue` During Instantiation

### ISSUE THREAD

**Issue by s0kil** [digitallyinduced/ihp#936]
`componentFromState @Counter (Counter { value = 100 })` should use the provided  initial state instead of relying on the `initialValue` defined in the Counter component module.

**Comment by MurakamiKennzo** [comment]
any updated here?

**Comment by mpscholten** [comment]
Not yet 👍

**Comment by MurakamiKennzo** [comment]
😭, What's the main problem here, I read some ssc code just now and think It's may not a tricky.

**Comment by fidel-ml** [comment]
What about something as this:
```
class Component state action configData | state -> action, state -> configData where
  initialState :: configData -> state
  ...
```
Then you may have both static and dynamic configuration data for the components (as you are able to choose the `configData`...).
For example
```
instance Component Counter CounterController Int where
  initialState n = Counter { value = n }
```

**Comment by mpscholten** [comment]
I like this idea. It's very close to how react.js works as well 👍

**Comment by s0kil** [comment]
With `class Component state action configData | state -> action, state -> configData where` it will require ConfigData to be provided on every instance of the component, correct? `component @Counter configData`, I can no longer do `component @Counter`, or am I misunderstanding something?

**Comment by mpscholten** [comment]
yep that's correct. If you set `configData = ()` you could use `component @Counter ()` as a workaround to get the old behaviour

**Comment by fidel-ml** [comment]
You may also provide a `defaultConfig`, and let `component` use it, and another `componentWith` for the new behaviour...

**Comment by mpscholten** [comment]
> You could maybe put the props as data attribute into the html and send it somehow with the first websocket call and set it before you change the state. But this is probably a stupid idea....

I think this might actually be a good option here.

The ControllerContext cannot be used as we're having two requests here. The first that renders the SSC `componentWithProps Counter { value = 100 }`. Then the JS connects via the WebSocket which is a second request.

We've had a similiar problem with IHP AutoRefresh (keeping state across multiple requests). AutoRefresh solves this by having a [server-global session store](https://github.com/digitallyinduced/ihp/blob/master/IHP/Server.hs#L53). On the first request it creates a new `AutoRefreshSession` data type and stores it into the session store. It also puts a unique ID of the `AutoRefreshSession` into the first response via [a meta tag](https://github.com/digitallyinduced/ihp/blob/master/IHP/AutoRefresh/View.hs#L12). The AutoRefresh JS now passes the ID onto the WebSocket. The WebSocket handler [looks up the correct session by the ID](https://github.com/digitallyinduced/ihp/blob/master/IHP/AutoRefresh.hs#L111). This way we share state across multiple requests. Passing the state around here explicitly via JSON doesn't work here as the state here is the ControllerContext which cannot be serialized as JSON.

**Comment by leobm** [comment]
Hello Marc @mpscholten ,

I was playing around a while ago. I now pass the props in the websocket URL on the first request and then set the state in the ComponentsController over it. 

My example runs with it. But I really only have rudimentary Haskell knowledge, so the code is probably really bad. I've somehow managed to get there with a lot of trial and error. Here is the branch with my changes.
https://github.com/leobm/ihp/tree/feature-component-props

I also wrote a test-componenet. 
https://github.com/leobm/ihp-test-component-with-props

but somehow I didn't manage to link to my feature branch in default.nix. I have included my feature branch for testing simply over an IHP subdirectory.

Both projects compile. I have only strangely in Visual Code (Haskell plugin) with the language server problems. Get then e.g. Web.Component.Counter File the following error displayed:
"- Expected kind '* -> Constraint',
    but 'Component Counter CounterController' has kind 'Constraint'.
- In the instance declaration for
    'Component Counter CounterController CounterProps'typecheck
"

do you know where the problem is?
